esda/gamma.py

```diff
@@ -238,6 +235,37 @@ def __calc(self, z, op):
                 zw = list(zip(neighbors, wijs))
                 zs[i] = sum([wij * op(z, i, j) for j, wij in zw])
             g = zs.sum()
+
+        return g
+
+    def __calc_g(self, z, op):
+        if op == "c":  # cross-product
+            zl = self.w.lag(z)
+            g = (z * zl).sum()
+        elif op == "s":  # squared difference
+            z = pd.Series(z, index=self.w.unique_ids)
+            z2 = z**2
+            focal = self.w._adjacency.index.get_level_values(0)
+            neighbour = self.w._adjacency.index.get_level_values(1)
+            g = (
+                self.w._adjacency.values
+                * (
+                    z2[focal].values
+                    - 2.0 * z[focal].values * z[neighbour].values
+                    + z2[neighbour].values
+                )
+            ).sum()
+        elif op == "a":  # absolute difference
+            z = pd.Series(z, index=self.w.unique_ids)
+            focal = self.w._adjacency.index.get_level_values(0)
+            neighbour = self.w._adjacency.index.get_level_values(1)
+            g = (
+                self.w._adjacency.values
+                * (np.abs(z[focal].values - z[neighbour].values))
+            ).sum()
+        else:  # any previously defined function op
+            raise NotImplementedError
```

Comment:
What is the expectation for a custom function here? It is undocumented. I see how it is defined in tests but is that the expectation? Is some downstream code relying on this?

Reply:
I don't think any downstream code relies on this.

This is handling the `operation` option that specifies the form of the attribute similarity measure, which is documented in the class docstring. But maybe you are referring to something else not being documented?

Reply:
I see the example but I wasn't sure what the expectations for the function signature are. I can derive that from the example but I'm a bit hesitant trying to replicate that as the function optimised for graph would probably look different.

Reply:
I don't think there were any expectations. More like a place holder in case other similarity functions were proposed.

On a related issue, would it be possible to add some helper methods on the graph to simplify these: https://github.com/pysal/esda/blob/9987fd4f63b2857fcfe465e22436913d0f54b10c/esda/gamma.py#L248-L249

Would that make sense?

Reply:
Yeah, I've been writing that way too often recently so it certainly deserves a helper method.